### PR TITLE
6432 - implement a first simple

### LINKED ISSUE #6432: Implement first (simple/PoC) version of a OIDC auth provider
This is part of epic #5974.

----

Since the refactoring of OAuth2 infrastructure in #5991 and #6364 it was made possible to introduce a new authentication mechanism using the [Nimbus OAuth 2.0 SDK with OpenID Connect extensions](https://connect2id.com/products/nimbus-oauth-openid-connect-sdk).
(ScribeJava has a long missing feature request for OIDC support, which has never been attacked).

This first implementation should enable us:
* load a configuration
* retrieve metadata
* login at the provider
* retrieve standardized claims like email, name, ...
* allow multiple providers so it can be reused for different IdPs without touching codebase

### PR DESCRIPTION
**Pull request by poikilotherm**
**What this PR does / why we need it**:
This pull request adds very basic support for authentication with an OpenID Connect capable identity provider.

We need this is as a basis to solve #5974, which is about using a more sophisticated user authentication system for Dataverse.

**Which issue(s) this PR closes**:
- Closes #6432

**Special notes for your reviewer**:
This has been tested so far with an internal IdP based on Unity IDM.
I think about introducing integration testing this feature by using [Testcontainers](https://www.testcontainers.org), but this seems like beyond scope for this PR and keeping that narrow.

**Suggestions on how to test this**:
We should test this with Google, maybe with ORCID.

**Does this PR introduce a user-facing change?**:
Kind of. It reuses all the existing OAuth2 UI stuff, but of course enables new buttons for providers. It might become an issue in case someone really wants to add a big bunch of stuff, as the list grows in that case.

It might affects the order of auth providers, too, as the sorting feature for the UI needed to be refactored.

**Is there a release notes update needed for this change?**
Sure. But maybe this should be a separate PR, as there are some more things still to come.

**Additional documentation**:
This PR is a demo for #6226 on @pdurbins [demand on IRC](http://irclog.iq.harvard.edu/dataverse/2019-12-04#i_113824)

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/LoginPage.java [hunk 1]
```diff
@@ -129,16 +125,20 @@ public void init() {
...
-                    infos.add(authenticationProvider.getInfo());
-                }
+        List<AuthenticationProvider> idps = new ArrayList<>(authSvc.getAuthenticationProviders());
+        
+        // sort by order first. in case of same order values, be deterministic in UI and sort by id, too.
+        Collections.sort(idps, Comparator.comparing(AuthenticationProvider::getOrder).thenComparing(AuthenticationProvider::getId));
+        
+        for (AuthenticationProvider idp : idps) {
+            if (idp != null) {
+                infos.add(idp.getInfo());
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/LoginPage.java]: Is the change above necessary for OIDC to work? I don't think so. It does look like a nice cleanup though. I like how it's deterministic. Perhaps this change should be documented.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java [hunk 2]
```diff
@@ -207,6 +214,14 @@ public AuthenticationProviderDisplayInfo getInfo() {
         return new AuthenticationProviderDisplayInfo(getId(), getTitle(), getSubTitle());
     }
     
+    /**
+     * Used in {@link LoginPage#listAuthenticationProviders()} for sorting the providers in the UI
+     * TODO: this might be extended to use a value set by the admin when configuring the provider via JSON.
+     * @return an integer value (sort ascending)
+     */
+    @Override
+    public int getOrder() { return 100; }
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java]: What if there are more than 100 authentication providers? Some have the same number? Is it ok to have two 42s?

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
+package edu.harvard.iq.dataverse.authorization.providers.oauth2.oidc;
+
+import com.github.scribejava.core.builder.api.DefaultApi20;
+import com.nimbusds.oauth2.sdk.*;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java]: In general, we try not to use star imports. I regret to say that we don't mention this at http://guides.dataverse.org/en/4.18.1/developers/coding-style.html 😄

**Comment by pdurbin** [comment]
> **Suggestions on how to test this**:
> We should test this with Google, maybe with ORCID.

@poikilotherm can you also provide an OIDC environment for testing the actual feature this pull request introduces?

File: src/main/java/edu/harvard/iq/dataverse/LoginPage.java (hunk 1 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/LoginPage.java]: Actually yes it is. From this function, `authSvc.getAuthenticationProviderIdsSorted()` was called, which is deleted below. This was done in 3fd326c23 and here's what I wrote in the commit log:

> Refactor LoginPage.listAuthenticationProviders() to represent actual provider configuration.
>
> Before this refactoring, a static list of providers has been used, no matter if they where
enabled or not. This is not acceptable for the new OIDC provider, as we might have multiple
of 'em and it retrieves metadata on creation.
>
> So the idea was to not create some static list of IDs (which had been used to look up the real
provider from the "registry"), but instead get the registred providers, sort them and
retrieve the display info.
>
> Sorting is done by a numerical value first, then by id attribute (which is set by factories),
so we have a deterministic, unchanging order for the user.
>
> This lead to the creation of a new method AuthenticationProvider.getOrder(), defaulting to 1
and overridden by implementations. This opens up the possibility to change ordering by
configuration (as suggested in an old TODO comment of the refactored code)

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java (hunk 2 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java]: Actually it doesn't matter if some have the same number, as they are sorted by number first, then by name.

This is mostly about sorting Builtin first (order = 1), Shib second (order=20) and OAuth last (all order = 100), as it has been sorted before my refactoring.

Once we have a configurable value for this, you could also realize sth. like provider A first, then all the others.
[Please see also my comment above.](https://github.com/IQSS/dataverse/pull/6433/files/511827429f280ee7ad997fa68fc85694a556640d?file-filters%5B%5D=.java&file-filters%5B%5D=.properties#r354178859)

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java (hunk 3 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java]: I'm sorry, IntelliJ IDEA does this by default.
Should I mention checkstyle again to catch this sort of stuff?
We could create a Jenkins job for it...

Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OIDCAuthProvider.java]: Oh and should I change this in a new PR as this one is already merged?

**Comment by poikilotherm** [comment]
I can report great success for using Google OpenID Connect support. Worked flawless, does what it should do. I love standards... :smile: 

Should we talk to @alejandratenorio and @Gerafp for testing it with Microsoft Azure AD?

**Pull request merged**